src/tls.rs

```diff
@@ -1,6 +1,354 @@
-//! TLS-related helper types
-use core::ffi::{c_char, CStr};
-use core::fmt::Debug;
+//! Type safe abstraction for esp-tls
+
+use core::{
+    ffi::CStr,
+    fmt::{Debug, Display},
+    time::Duration,
+};
+
+use embedded_svc::io;
+
+use crate::{
+    errors::EspIOError,
+    sys::{EspError, ESP_ERR_INVALID_SIZE, ESP_ERR_NO_MEM, ESP_FAIL},
+};
+
+/// see https://www.ietf.org/rfc/rfc3280.txt ub-common-name-length
+const MAX_COMMON_NAME_LENGTH: usize = 64;
+
+/// Wrapper for `esp-tls` module. Only supports synchronous operation for now.
+pub struct EspTls {
+    reader: EspTlsRead,
+    writer: EspTlsWrite,
+}
+
+impl EspTls {
+    /// Create a new blocking TLS/SSL connection.
+    ///
+    /// This function establishes a TLS/SSL connection with the specified host in a blocking manner.
+    ///
+    /// # Errors
+    ///
+    /// * `ESP_ERR_INVALID_SIZE` if `cfg.alpn_protos` exceeds 9 elements or avg 10 bytes/ALPN
+    /// * `ESP_ERR_NO_MEM` if TLS context could not be allocated
+    /// * `ESP_FAIL` if connection could not be established
+    pub fn new(host: &str, port: u16, cfg: &Config) -> Result<Self, EspError> {
+        let mut rcfg: esp_idf_sys::esp_tls_cfg = unsafe { std::mem::zeroed() };
```

Comment:
`crate::sys` - here and everywhere below. In fact, I am surprised how the `esp_idf_sys::` code even compiles, given that latest `master` is not even directly depending on `esp-idf-sys` anymore?

Reply:
UPDATE: Yeah, it is not, CI failed.